lib/webauthn/authenticator_response.rb

```diff
@@ -60,11 +73,25 @@ def valid_challenge?(original_challenge)
       WebAuthn::SecurityUtils.secure_compare(Base64.urlsafe_decode64(client_data.challenge), original_challenge)
     end
 
-    def valid_origin?(original_origin)
-      client_data.origin == original_origin
+    # If the caller specifies the current origin, check that it exactly matches the signed origin.
+    def valid_current_origin?(current_origin)
+      client_data.origin == current_origin
+    end
+
+    def valid_origin?(rp_id)
+      # Per https://www.w3.org/TR/webauthn/#rp-id
+      # - The RP ID must be equal to the origin's effective domain, or a registrable domain suffix of the origin's effective domain.
+      # - The origin's scheme must be https.
+      # - The origin's port is unrestricted.
+      origin = URI.parse(client_data.origin)
+      # Calculate "registrable domain suffix" based on
+      # https://html.spec.whatwg.org/multipage/origin.html#is-a-registrable-domain-suffix-of-or-is-equal-to
+      valid_scope = origin.domain == rp_id || origin.domain.end_with?("." + rp_id)
+      valid_scope && origin.scheme == "https"
     end
 
     def valid_rp_id?(rp_id)
+      # TODO: rp_id cannot be a public suffix
```

Comment:
Ideally, this line would include a check against a public suffix library. I shied away from putting that in the initial PR.